Update the storage readme with concepts (#1005)
Azure/azure-sdk-for-c: sdk/docs/storage/README.md

* Upload blobs

## Getting Started

The Azure Storage Blobs Client library facilitates connectivity to the Azure Storage Blobs service. Since this service uses HTTP, you must plug in an HTTP adapter capable of sending the HTTP request to your device's HTTP stack. For information on how to do this, please see [this page][docs_platform_readme].

### Docs

For API documentation, please see the doxygen generated docs [here][azure_sdk_for_c_doxygen_docs]. You can find the Storage Blobs specific docs by navigating to the **Files -> File List** section near the top and choosing any of the header files prefixed with `az_storage_blobs`.

### Build

The Azure Storage Blobs library is compiled following the same steps listed on the root [README][azure_sdk_for_c_readme] documentation, under ["Getting Started Using the SDK"][azure_sdk_for_c_readme_getting_started].

The library targets made available via CMake are the following:

- `az::storage::blobs` - For Azure Storage Blobs features ([API documentation here][azure_sdk_for_c_doxygen_docs])

### Samples

This [page][samples_storage_blobs_readme] explains samples for the Azure Embedded C SDK Storage Blobs Client.

> 

Reviewer: You should consider using the markdown linter: https://github.com/Azure/azure-sdk-for-c/issues/459, https://github.com/Azure/azure-sdk-for-c/pull/959

nit: Remove extra new line.

Use the Azure SDK Storage Blobs Library for Embedded C to work with Azure Storage Blobs.

> * Upload blobs

Reviewer: This list just appears and the preamble sentence above it doesn't lead up to it. Reword?

## Getting Started

> The Azure Storage Blobs Client library is created to facilitate connectivity to Azure Storage Blobs services using the HTTP stack of the user's choice. This means that this SDK is **NOT** a platform but instead is a true SDK library.

Reviewer: I don't understand what we are trying to say with that last sentence. Seems odd to bring up.

```suggestion
The Azure Storage Blobs Client library is created to facilitate connectivity to Azure Storage Blobs services using the HTTP stack of the user's choice.
```

Author: It helps the user understand that they can control the underlying behavior.  That is distinct difference from the previous C SDKs which had implementations which could not be changed

Reviewer: Makes sense. I don't know how to word that well, but using overloaded terms like "it is not a platform", and "true SDK library" are confusing. I start asking these questions: What's a platform in this context? What's a true SDK library supposed to be?

Maybe...
```suggestion
The Azure Storage Blobs Client library is created to facilitate connectivity to Azure Storage Blobs services using the HTTP stack of the user's choice. This capability differentiates the Embedded SDK from the previous C SDKs which had underlying implementations that the user could not be changed.
```

> underlying transport (platform) 

Also, we have been a few conversations (with @JeffreyRichter) about terminology around platform, and HTTP transport. Let's make sure to keep those two distinct (i.e. platform != http transport adapter)
See this issue: https://github.com/Azure/azure-sdk-for-c/issues/935

Reviewer 2: I suggest this:
The Azure Storage Blobs Client library facilitates connectivity to the Azure Storage Blobs service. Since this service uses HTTP, you must plug in an HTTP adapter capable of sending the HTTP request to your device's HTTP stack. For information on how to do this, please see (link goes here).

(I'm assuming that we'll have more HTTP clients in the future and so the HTTP adapter stuff should be in 1 place and then simply referred to by the various clients)

### Docs

For API documentation, please see the doxygen generated docs [here][azure_sdk_for_c_doxygen_docs]. You can find the Storage Blobs specific docs by navigating to the **Files -> File List** section near the top and choosing any of the header files prefixed with `az_storage_blobs`.

### Build

The Azure Storage Blobs library is compiled following the same steps listed on the root [README][azure_sdk_for_c_readme] documentation, under ["Getting Started Using the SDK"][azure_sdk_for_c_readme_getting_started].

The library targets made available via CMake are the following:

> - `az::storage::blobs` - For Azure Storage Blobs features ([API documentation here][azure_sdk_for_c_doxygen_docs])

Reviewer: Why are we linking to our root README for azure storage blobs features? What section (if any) is relevant to the user here?

Author: Its not linking to the root readme.  It points to the root of the docs.  Unfortunately there isn't a way to reference into the docs w/o placing version specific information into the paths.  This is the best we can do for the time.

Reviewer: Oh I see. I clicked the wrong link.

I am assuming your intended link would be this (if we didn't have the version issue), correct?
https://azuresdkdocs.blob.core.windows.net/$web/c/docs/1.0.0-preview.3/az__storage__blobs_8h.html

### Samples

This [page][samples_storage_blobs_readme] explains samples for the Azure Embedded C SDK Storage Blobs Client.

### Prerequisites

For compiling the Azure SDK for Embedded C for the most common platforms (Windows and Linux), no further prerequisites are necessary.
Please follow the instructions in the [Getting Started](#Getting-Started) section above.
For compiling for specific target devices, please refer to their specific toolchain documentation.

## Key Concepts

### Authentication

The embedded C SDK currently supports SAS based authentication.  See [this page][storage_access_control_sas] for information on creating SAS tokens.
The client credential should be set to `AZ_CREDENTIAL_ANONYMOUS` when using SAS tokens.

### Creating the Storage Client

> To use the storage client the first action is to initialize the client with `az_storage_blobs_blob_client_init`.

Reviewer: ```suggestion
To use the storage client, the first action is to initialize the client with `az_storage_blobs_blob_client_init`.
```

```C
  az_storage_blobs_blob_client client;
  az_storage_blobs_blob_client_options options = az_storage_blobs_blob_client_options_default();

  az_storage_blobs_blob_client_init(

>           &client, az_span_from_str(getenv(URI_ENV)), AZ_CREDENTIAL_ANONYMOUS, &options);

Reviewer: We should just pass in `NULL` for options, to keep the getting started simple and "one-liner".

Author: I prefer using the explicit value.  It makes it explicit what we are doing.

Reviewer: I think not allowing `NULL`, as in keeping the options, optional, is a usability concern.
We have the pattern of passing in `NULL` to mean `default` in a whole bunch of existing SDK APIs. I think we should make sure both work, but that is related to the API and not docs (unrelated to this PR).

But you are right, leaving the option in the code snippet in the docs and being explicit is reasonable.

Remove the `NULL` parameter below in `az_storage_blobs_blob_upload`?

Author: We can't remove that NULL parameter.  That is the options for upload which are optional

Reviewer: By remove, I mean state that parameter explicitly. The same rationale you are applying here.

```

### Uploading a blob

Once the client is created it can be used to upload blobs.
```C

>   az_result const blob_upload_result = az_storage_blobs_blob_upload(
>       &client, &az_context_app, content_to_upload, NULL, &http_response)

Reviewer: ```suggestion
  az_result const blob_upload_result = az_storage_blobs_blob_upload(
      &client, &az_context_application, content_to_upload, NULL, &http_response)
```

Reviewer: Do we want to mention here a little bit about context and http response (maybe link out), or is the intention to detail this further in samples? I was wondering where the response came from.

Reviewer: FYI @antkmsft, keep an eye out for merge conflicts here related to https://github.com/Azure/azure-sdk-for-c/pull/988.

Azure SDK for Embedded C is licensed under the [MIT][azure_sdk_for_c_license] license.

<!-- LINKS -->

> [azure_pattern_circuit_breaker]: https://docs.microsoft.com/azure/architecture/patterns/circuit-breaker

Reviewer 2: I would remove any references to "circuit breaker pattern" from this document as our SDK does not have support for it built in.

Author: We don't have full circuit breakers but we do have some aspects with retry.  Also the article is a good read for handling network error patterns
